cli_tools/gce_windows_upgrade/upgrade_script.ps1

```diff
@@ -12,21 +12,50 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
+function Get-MetadataValue {
+  param (
+    [parameter(Mandatory=$true)]
+    [string]$key
+  )
+
+  # Returns the provided metadata value for a given key.
+  $url = "http://metadata.google.internal/computeMetadata/v1/instance/attributes/$key"
+  $max_attemps = 5
+  for ($i=0; $i -le $max_attemps; $i++) {
+    try {
+      $client = New-Object Net.WebClient
+      $client.Headers.Add('Metadata-Flavor', 'Google')
+      $value = ($client.DownloadString($url)).Trim()
+      Write-Host "Retrieved metadata for key $key with value $value."
+      return $value
+    }
+    catch [System.Net.WebException] {
+      # Sleep after each failure with no default value to give the network adapters time to become functional.
+      Start-Sleep -s 1
+    }
+  }
+  throw "Failed $max_attemps times to retrieve value from metadata for $key."
+}
+
 $script:install_media_drive = ''
 
 try {
   Write-Host 'GCEMetadataScripts: Beginning upgrade startup script.'
 
+  $script:expected_current_version = Get-MetadataValue -key 'expected-current-version'
+  $script:expected_new_version = Get-MetadataValue -key 'expected-new-version'
+  $script:install_folder = Get-MetadataValue -key 'install-folder'
+
   # Cleanup garbage files left by the previous failed upgrade to unblock a new upgrade.
   Remove-Item 'C:\$WINDOWS.~BT' -Recurse -ErrorAction SilentlyContinue
 
-  $ver=[System.Environment]::OSVersion.Version
-  if ("$($ver.Major).$($ver.Minor)" -ne "6.1") {
-    if ("$($ver.Major).$($ver.Minor)" -eq "6.3") {
-      Write-Host "GCEMetadataScripts: The instance is already running Windows 2012R2!"
+  $ver=(Get-ItemProperty "HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion").ProductName
+  if ($ver -ne $script:expected_current_version) {
+    if ($ver -eq $script:expected_new_version) {
```

Comment:
It's not considered since previously we trust that the "license" matches the os version. As you said, to be more secure, we should exclude any unexpected version here as well. Let me add that logic.
We don't use "<" or ">" here since version may not be simply higher or lower, it can just be unexpected. For example, windows 11 vs windows server 2022.